### ISSUE TITLE
Can't launch OAS: VRSettingError_UnsetSettingHasNoDefault (Windows passing null audio device)

### ISSUE THREAD

**Issue by SpXxwW** [OpenVR-Advanced-Settings/OpenVR-AdvancedSettings#357]
**Describe the bug**
OAS doesn't launch. 

...
[INFO] 2020-01-27 12:16:40: VR Runtime Path: C:/Program Files (x86)/Steam/steamapps/common/SteamVR
[ERROR] 2020-01-27 12:16:41: Could not get recording override setting: VRSettingsError_UnsetSettingHasNoDefault
[ERROR] 2020-01-27 12:16:41: Could not get playback override setting: VRSettingsError_UnsetSettingHasNoDefault

I have reinstalled SteamVR (and tried SteamVR beta) and cleared all settings and reinstalled OAS multiple times.

**To Reproduce**
1. Launch SteamVR

**Expected behavior**
OAS should launch with SteamVR.

**Screenshots**
No screenshots.

**Software (please complete the following information):**
 - OS: Windows 10
 - Advanced Settings Version: 4.0.1

**Hardware (please complete the following information):**
 - CPU: i7 7700k
 - Graphics Card: Radeon RX 480
 - HMD: HTC Vive (Index controllers)

**Additional context**
I have done alot of googling, but there seems to be no information about the issue.

**Comment by username223** [comment]
Thanks for reporting.

The
```
[ERROR] 2020-01-27 12:16:41: Could not get recording override setting: VRSettingsError_UnsetSettingHasNoDefault
[ERROR] 2020-01-27 12:16:41: Could not get playback override setting: VRSettingsError_UnsetSettingHasNoDefault
```
is a SteamVR issue, but OVRAS should still launch normally.

Could you paste the full log from `%appdata%\AdvancedSettings-Team\AdvancedSettings.log`?

**Comment by SpXxwW** [comment]
[INFO] 2020-01-27 19:45:39: Application started (Version 4.0.1)
[INFO] 2020-01-27 19:45:39: Log File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/AdvancedSettings.log
[INFO] 2020-01-27 19:45:39: Command line arguments processed.
[INFO] 2020-01-27 19:45:39: OpenVR initialized successfully.
[INFO] 2020-01-27 19:45:39: Settings File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/OpenVRAdvancedSettings.ini
[INFO] 2020-01-27 19:45:39: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/OpenVRAdvancedSettings.ini
[INFO] 2020-01-27 19:45:39: VR Runtime Path: C:/Program Files (x86)/Steam/steamapps/common/SteamVR
[ERROR] 2020-01-27 19:45:40: Could not get recording override setting: VRSettingsError_UnsetSettingHasNoDefault
[ERROR] 2020-01-27 19:45:40: Could not get playback override setting: VRSettingsError_UnsetSettingHasNoDefault

This is everything what happens per launch. I can't see any difference in VR menu, so even the OAS button does not appear.

**Comment by username223** [comment]
That's strange.

Could you try installing and running [this](https://ci.appveyor.com/api/buildjobs/tx0rkl0qii5sc75g/artifacts/installer%2FAdvancedSettings-4.0.1-dev%2B50c84156-Installer.exe) build instead?

**Comment by SpXxwW** [comment]
```
[INFO] 2020-01-28 12:51:44: Application started (Version 4.0.1-dev+50c84156)
[INFO] 2020-01-28 12:51:44: Log File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/AdvancedSettings.log
[INFO] 2020-01-28 12:51:44: Settings File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/OpenVRAdvancedSettings.ini
[INFO] 2020-01-28 12:51:44: defaultlockXToggle: 'false' | lockYToggle: 'false' | lockZToggle: 'false' | momentumSave: 'false' | turnBindLeft: 'false' | turnBindRight: 'false' | turnBounds: 'false' | moveShortcutLeft: 'false' | moveShortcutRight: 'false' | dragBounds: 'false' | allowExternalEdits: 'false' | oldStyleMotion: 'false' | universeCenteredRotation: 'false' | enableSeatedMotion: 'false' | adjustChaperone: 'true' | showLogMatricesButton: 'false' | simpleRecenter: 'false' | disableVersionCheck: 'false' | previousShutdownSafe: 'true' | vsyncDisabled: 'false' | crashRecoveryDisabled: 'false' | enableDebug: 'false' | pttEnabled: 'false' | pttShowNotification: 'false' | micProximitySensorCanMute: 'false' | micReversePtt: 'false' | alarmEnabled: 'false' | alarmIsModal: 'true' | vrcDebug: 'false' | brightnessEnabled: 'false' | isOverlayMethodActive: 'false' | colorOverlayEnabled: 'false' | chaperoneSwitchToBeginnerEnabled: 'false' | chaperoneHapticFeedbackEnabled: 'false' | chaperoneAlarmSoundEnabled: 'false' | chaperoneAlarmSoundLooping: 'true' | chaperoneAlarmSoundAdjustVolume: 'false' | chaperoneShowDashboardEnabled: 'false' | disableChaperone: 'false' | defaultheightToggleOffset: '-1.000000' | gravityStrength: '9.800000' | flingStrength: '1.000000' | brightnessOpacityValue: '0.000000' | colorOverlayOpacity: '0.000000' | colorRedNew: '1.000000' | colorGreenNew: '1.000000' | colorBlueNew: '1.000000' | chaperoneSwitchToBeginnerDistance: '0.500000' | chaperoneHapticFeedbackDistance: '0.500000' | chaperoneAlarmSoundDistance: '0.500000' | chaperoneShowDashboardDistance: '0.500000' | fadeDistanceRemembered: '0.500000' | defaultkeyboardOne: '^>m' | keyboardTwo: '^>m' | keyboardThree: '^>m' | defaultsnapTurnAngle: '4500' | smoothTurnRate: '100' | dragComfortFactor: '0' | turnComfortFactor: '0' | frictionPercent: '0' | debugState: '0' | customTickRateMs: '20' | alarmHour: '0' | alarmMinute: '0' | 
[INFO] 2020-01-28 12:51:44: Command line arguments processed.
[INFO] 2020-01-28 12:51:44: OpenVR initialized successfully.
[INFO] 2020-01-28 12:51:44: VR Runtime Path: C:/Program Files (x86)/Steam/steamapps/common/SteamVR
[ERROR] 2020-01-28 12:51:45: Could not get recording override setting: VRSettingsError_UnsetSettingHasNoDefault
[ERROR] 2020-01-28 12:51:45: Could not get playback override setting: VRSettingsError_UnsetSettingHasNoDefault

[INFO] 2020-01-28 12:53:58: Application started (Version 4.0.1-dev+50c84156)
[INFO] 2020-01-28 12:53:58: Log File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/AdvancedSettings.log
[INFO] 2020-01-28 12:53:58: Settings File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/OpenVRAdvancedSettings.ini
[INFO] 2020-01-28 12:53:58: defaultlockXToggle: 'false' | lockYToggle: 'false' | lockZToggle: 'false' | momentumSave: 'false' | turnBindLeft: 'false' | turnBindRight: 'false' | turnBounds: 'false' | moveShortcutLeft: 'false' | moveShortcutRight: 'false' | dragBounds: 'false' | allowExternalEdits: 'false' | oldStyleMotion: 'false' | universeCenteredRotation: 'false' | enableSeatedMotion: 'false' | adjustChaperone: 'true' | showLogMatricesButton: 'false' | simpleRecenter: 'false' | disableVersionCheck: 'false' | previousShutdownSafe: 'true' | vsyncDisabled: 'false' | crashRecoveryDisabled: 'false' | enableDebug: 'false' | pttEnabled: 'false' | pttShowNotification: 'false' | micProximitySensorCanMute: 'false' | micReversePtt: 'false' | alarmEnabled: 'false' | alarmIsModal: 'true' | vrcDebug: 'false' | brightnessEnabled: 'false' | isOverlayMethodActive: 'false' | colorOverlayEnabled: 'false' | chaperoneSwitchToBeginnerEnabled: 'false' | chaperoneHapticFeedbackEnabled: 'false' | chaperoneAlarmSoundEnabled: 'false' | chaperoneAlarmSoundLooping: 'true' | chaperoneAlarmSoundAdjustVolume: 'false' | chaperoneShowDashboardEnabled: 'false' | disableChaperone: 'false' | defaultheightToggleOffset: '-1.000000' | gravityStrength: '9.800000' | flingStrength: '1.000000' | brightnessOpacityValue: '0.000000' | colorOverlayOpacity: '0.000000' | colorRedNew: '1.000000' | colorGreenNew: '1.000000' | colorBlueNew: '1.000000' | chaperoneSwitchToBeginnerDistance: '0.500000' | chaperoneHapticFeedbackDistance: '0.500000' | chaperoneAlarmSoundDistance: '0.500000' | chaperoneShowDashboardDistance: '0.500000' | fadeDistanceRemembered: '0.500000' | defaultkeyboardOne: '^>m' | keyboardTwo: '^>m' | keyboardThree: '^>m' | defaultsnapTurnAngle: '4500' | smoothTurnRate: '100' | dragComfortFactor: '0' | turnComfortFactor: '0' | frictionPercent: '0' | debugState: '0' | customTickRateMs: '20' | alarmHour: '0' | alarmMinute: '0' | 
[INFO] 2020-01-28 12:53:58: Desktop mode enabled.
[INFO] 2020-01-28 12:53:58: Sound effects disabled.
[INFO] 2020-01-28 12:53:58: Command line arguments processed.
[INFO] 2020-01-28 12:54:01: OpenVR initialized successfully.
[INFO] 2020-01-28 12:54:01: VR Runtime Path: C:/Program Files (x86)/Steam/steamapps/common/SteamVR
[ERROR] 2020-01-28 12:54:02: Could not get recording override setting: VRSettingsError_UnsetSettingHasNoDefault
[ERROR] 2020-01-28 12:54:02: Could not get playback override setting: VRSettingsError_UnsetSettingHasNoDefault
```

Same thing happens here. I will provide any information you might need, just ask. :)

**Comment by username223** [comment]
Alright, could you
* Go to `%localappdata%/crashdumps` and delete all files containing `AdvancedSettings.exe` before running the next build
* Download [this](https://ci.appveyor.com/api/buildjobs/e6n7bt8053hqem80/artifacts/installer%2FAdvancedSettings-4.0.1-dev%2B2c58b4a3-Installer.exe) build
* Run it
* Upload logs like before and
* Upload the `AdvancedSettings.exe.*.dmp` file from `%localappdata%/crashdumps`

**Comment by SpXxwW** [comment]
```
[INFO] 2020-01-28 13:23:50: Application started (Version 4.0.1-dev+50c84156)
[INFO] 2020-01-28 13:23:50: Log File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/AdvancedSettings.log
[INFO] 2020-01-28 13:23:50: Settings File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/OpenVRAdvancedSettings.ini
[INFO] 2020-01-28 13:23:50: defaultlockXToggle: 'false' | lockYToggle: 'false' | lockZToggle: 'false' | momentumSave: 'false' | turnBindLeft: 'false' | turnBindRight: 'false' | turnBounds: 'false' | moveShortcutLeft: 'false' | moveShortcutRight: 'false' | dragBounds: 'false' | allowExternalEdits: 'false' | oldStyleMotion: 'false' | universeCenteredRotation: 'false' | enableSeatedMotion: 'false' | adjustChaperone: 'true' | showLogMatricesButton: 'false' | simpleRecenter: 'false' | disableVersionCheck: 'false' | previousShutdownSafe: 'true' | vsyncDisabled: 'false' | crashRecoveryDisabled: 'false' | enableDebug: 'false' | pttEnabled: 'false' | pttShowNotification: 'false' | micProximitySensorCanMute: 'false' | micReversePtt: 'false' | alarmEnabled: 'false' | alarmIsModal: 'true' | vrcDebug: 'false' | brightnessEnabled: 'false' | isOverlayMethodActive: 'false' | colorOverlayEnabled: 'false' | chaperoneSwitchToBeginnerEnabled: 'false' | chaperoneHapticFeedbackEnabled: 'false' | chaperoneAlarmSoundEnabled: 'false' | chaperoneAlarmSoundLooping: 'true' | chaperoneAlarmSoundAdjustVolume: 'false' | chaperoneShowDashboardEnabled: 'false' | disableChaperone: 'false' | defaultheightToggleOffset: '-1.000000' | gravityStrength: '9.800000' | flingStrength: '1.000000' | brightnessOpacityValue: '0.000000' | colorOverlayOpacity: '0.000000' | colorRedNew: '1.000000' | colorGreenNew: '1.000000' | colorBlueNew: '1.000000' | chaperoneSwitchToBeginnerDistance: '0.500000' | chaperoneHapticFeedbackDistance: '0.500000' | chaperoneAlarmSoundDistance: '0.500000' | chaperoneShowDashboardDistance: '0.500000' | fadeDistanceRemembered: '0.500000' | defaultkeyboardOne: '^>m' | keyboardTwo: '^>m' | keyboardThree: '^>m' | defaultsnapTurnAngle: '4500' | smoothTurnRate: '100' | dragComfortFactor: '0' | turnComfortFactor: '0' | frictionPercent: '0' | debugState: '0' | customTickRateMs: '20' | alarmHour: '0' | alarmMinute: '0' | 
[INFO] 2020-01-28 13:23:50: Forcing removal of applications manifest.
[INFO] 2020-01-28 13:23:50: Command line arguments processed.
[INFO] 2020-01-28 13:23:51: OpenVR initialized successfully.
[INFO] 2020-01-28 13:23:51: Attempting to Remove Manifest At:  C:\Program Files\OpenVR-AdvancedSettings\manifest.vrmanifest
[INFO] 2020-01-28 13:23:51: Manifest Removed Successfully
[INFO] 2020-01-28 13:23:51: Manifest removed.
[INFO] 2020-01-28 13:24:31: Application started (Version 4.0.1-dev+2c58b4a3)
[INFO] 2020-01-28 13:24:31: Log File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/AdvancedSettings.log
[INFO] 2020-01-28 13:24:31: Settings File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/OpenVRAdvancedSettings.ini
[INFO] 2020-01-28 13:24:31: defaultlockXToggle: 'false' | lockYToggle: 'false' | lockZToggle: 'false' | momentumSave: 'false' | turnBindLeft: 'false' | turnBindRight: 'false' | turnBounds: 'false' | moveShortcutLeft: 'false' | moveShortcutRight: 'false' | dragBounds: 'false' | allowExternalEdits: 'false' | oldStyleMotion: 'false' | universeCenteredRotation: 'false' | enableSeatedMotion: 'false' | adjustChaperone: 'true' | showLogMatricesButton: 'false' | simpleRecenter: 'false' | disableVersionCheck: 'false' | previousShutdownSafe: 'true' | vsyncDisabled: 'false' | crashRecoveryDisabled: 'false' | enableDebug: 'false' | pttEnabled: 'false' | pttShowNotification: 'false' | micProximitySensorCanMute: 'false' | micReversePtt: 'false' | alarmEnabled: 'false' | alarmIsModal: 'true' | vrcDebug: 'false' | brightnessEnabled: 'false' | isOverlayMethodActive: 'false' | colorOverlayEnabled: 'false' | chaperoneSwitchToBeginnerEnabled: 'false' | chaperoneHapticFeedbackEnabled: 'false' | chaperoneAlarmSoundEnabled: 'false' | chaperoneAlarmSoundLooping: 'true' | chaperoneAlarmSoundAdjustVolume: 'false' | chaperoneShowDashboardEnabled: 'false' | disableChaperone: 'false' | defaultheightToggleOffset: '-1.000000' | gravityStrength: '9.800000' | flingStrength: '1.000000' | brightnessOpacityValue: '0.000000' | colorOverlayOpacity: '0.000000' | colorRedNew: '1.000000' | colorGreenNew: '1.000000' | colorBlueNew: '1.000000' | chaperoneSwitchToBeginnerDistance: '0.500000' | chaperoneHapticFeedbackDistance: '0.500000' | chaperoneAlarmSoundDistance: '0.500000' | chaperoneShowDashboardDistance: '0.500000' | fadeDistanceRemembered: '0.500000' | defaultkeyboardOne: '^>m' | keyboardTwo: '^>m' | keyboardThree: '^>m' | defaultsnapTurnAngle: '4500' | smoothTurnRate: '100' | dragComfortFactor: '0' | turnComfortFactor: '0' | frictionPercent: '0' | debugState: '0' | customTickRateMs: '20' | alarmHour: '0' | alarmMinute: '0' | 
[INFO] 2020-01-28 13:24:31: Forcing install of applications manifest.
[INFO] 2020-01-28 13:24:31: Command line arguments processed.
[INFO] 2020-01-28 13:24:31: OpenVR initialized successfully.
[INFO] 2020-01-28 13:24:31: Manifest Installed at: C:\Program Files\OpenVR-AdvancedSettings\manifest.vrmanifest
[INFO] 2020-01-28 13:24:31: Manifest reinstalled.
[INFO] 2020-01-28 13:25:07: Application started (Version 4.0.1-dev+2c58b4a3)
[INFO] 2020-01-28 13:25:07: Log File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/AdvancedSettings.log
[INFO] 2020-01-28 13:25:07: Settings File: C:/Users/SpXxw/AppData/Roaming/AdvancedSettings-Team/OpenVRAdvancedSettings.ini
[INFO] 2020-01-28 13:25:07: defaultlockXToggle: 'false' | lockYToggle: 'false' | lockZToggle: 'false' | momentumSave: 'false' | turnBindLeft: 'false' | turnBindRight: 'false' | turnBounds: 'false' | moveShortcutLeft: 'false' | moveShortcutRight: 'false' | dragBounds: 'false' | allowExternalEdits: 'false' | oldStyleMotion: 'false' | universeCenteredRotation: 'false' | enableSeatedMotion: 'false' | adjustChaperone: 'true' | showLogMatricesButton: 'false' | simpleRecenter: 'false' | disableVersionCheck: 'false' | previousShutdownSafe: 'true' | vsyncDisabled: 'false' | crashRecoveryDisabled: 'false' | enableDebug: 'false' | pttEnabled: 'false' | pttShowNotification: 'false' | micProximitySensorCanMute: 'false' | micReversePtt: 'false' | alarmEnabled: 'false' | alarmIsModal: 'true' | vrcDebug: 'false' | brightnessEnabled: 'false' | isOverlayMethodActive: 'false' | colorOverlayEnabled: 'false' | chaperoneSwitchToBeginnerEnabled: 'false' | chaperoneHapticFeedbackEnabled: 'false' | chaperoneAlarmSoundEnabled: 'false' | chaperoneAlarmSoundLooping: 'true' | chaperoneAlarmSoundAdjustVolume: 'false' | chaperoneShowDashboardEnabled: 'false' | disableChaperone: 'false' | defaultheightToggleOffset: '-1.000000' | gravityStrength: '9.800000' | flingStrength: '1.000000' | brightnessOpacityValue: '0.000000' | colorOverlayOpacity: '0.000000' | colorRedNew: '1.000000' | colorGreenNew: '1.000000' | colorBlueNew: '1.000000' | chaperoneSwitchToBeginnerDistance: '0.500000' | chaperoneHapticFeedbackDistance: '0.500000' | chaperoneAlarmSoundDistance: '0.500000' | chaperoneShowDashboardDistance: '0.500000' | fadeDistanceRemembered: '0.500000' | defaultkeyboardOne: '^>m' | keyboardTwo: '^>m' | keyboardThree: '^>m' | defaultsnapTurnAngle: '4500' | smoothTurnRate: '100' | dragComfortFactor: '0' | turnComfortFactor: '0' | frictionPercent: '0' | debugState: '0' | customTickRateMs: '20' | alarmHour: '0' | alarmMinute: '0' | 
[INFO] 2020-01-28 13:25:07: Command line arguments processed.
[INFO] 2020-01-28 13:25:10: OpenVR initialized successfully.
[INFO] 2020-01-28 13:25:10: VR Runtime Path: C:/Program Files (x86)/Steam/steamapps/common/SteamVR
[DEBUG] 2020-01-28 13:25:10: Pre stage 1 init.
[ERROR] 2020-01-28 13:25:10: Could not get recording override setting: VRSettingsError_UnsetSettingHasNoDefault
[ERROR] 2020-01-28 13:25:10: Could not get playback override setting: VRSettingsError_UnsetSettingHasNoDefault
```
Logs of me reinstalling OAS and then launching it. 

[AdvancedSettings.exe.17936.dmp.zip](https://github.com/OpenVR-Advanced-Settings/OpenVR-AdvancedSettings/files/4122096/AdvancedSettings.exe.17936.dmp.zip)

**Comment by username223** [comment]
Alright, I believe I've found the problem.

Could you tell me which language Windows displays, which localization Windows uses (which format for dates, numbers, etc.) and the names of all your audio devices (speaker and microphone)?

**Comment by SpXxwW** [comment]
Windows display language: English
Windows localization: Finnish

Speakers:
1. XG270HU (this is my main monitor)
2. SAMSUNG (this is TV my graphics card is plugged in)
3. Speakers (HTC Vive)
4. Speakers, Realtek High Definition Audio (These are normal speakers)
5. Headset Earphone, Sennheiser Communication Audio (My sennheiser USB audiocard)
6. Speakers, Sennheiser Main Audio (These are the sennheiser earphone speakers)

Mics:
1. Microphone (HTC Vive)
2. Heahset Microphone, Sennheiser Communication Audio

**Comment by username223** [comment]
Alright, I should have a fix in [this](https://ci.appveyor.com/api/buildjobs/8kqs3h56x4850se1/artifacts/installer%2FAdvancedSettings-4.0.1-dev%2B0b2cacd2-Installer.exe) build.

____

The previous dump file had a nullpointer exception in `std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>>.to_bytes()` because Windows passed null from `pProps->GetValue( PKEY_Device_FriendlyName, &varString )` in our `AudioManagerWindows::getDeviceName( IMMDevice* device )` function.

**Comment by SpXxwW** [comment]
That works! Thank you! I can see the null device in speakers drop down menu as empty item, but that should be fine (as long as I don't touch that :D). I can't test any functionality right now, but I assume everything works fine.

**Comment by username223** [comment]
Alright, cool. Do you have any idea what the name of the offending device is?

The fix will be in the next release build, but until then you can just use the previous build that works for you. This issue will be closed when the fix has been merged into the master branch.

**Comment by SpXxwW** [comment]
I listed 6 audio devices and OAS listed 8. (1 is Realtek Digital Output) and the last unlisted is the null device. I don't have any other audio device plugged or device manager or sound control panel does not show anything that I have not mentioned before. There was an issue with my Vive some time ago where windows corrupted all its drivers and with some trial and error I were able to make windows recognize Vive again. But maybe that left some corrupted audio drivers behind? Other than that I have no idea. I might try googling later to find out the problematic device information.
